docs/src/apalache/fold.md

```diff
@@ -0,0 +1,213 @@
+
+```
+CONSTANT negInf \* Max(<<>>) = -inf, but TLA+ has no literal for ±inf, so we use it as a constant
+
+RECURSIVE MaxRec(_)
+MaxRec(seq) == IF seq = <<>>
+               THEN negInf
+               ELSE LET tailMax == MaxRec(Tail(seq))
+                    IN IF tailMax > Head(seq)
+                       THEN tailMax
+                       ELSE Head(seq)
+
+MaxFold(seq) == LET Max(p,q) == IF p > q THEN p ELSE q
+                IN FoldSeq( Max, negInf, seq )
+```
+
+The first advantage of the fold implementation, we feel, is that it is much more clear and concise. It also does not require a termination condition, unlike the recursive case.
+One inherent problem of using recursive operators with a symbolic encoding, is the inability to estimate termination.
+While it may be immediately obvious to a human, that `MaxRec` terminates after no more than `Len(seq)` steps, automatic termination analysis is, in general, a rather complex form of static analysis.
+Apalache addresses this by finitely unrolling recursive operators and requires users to provide unroll limits (`UNROLL_LIMIT_MaxRec == ...`), which serve as a static upper bound to the number of recursive re-entries, because in general, recursive operators may take an unpredictable number of steps (e.g. computing the Collatz sequence) or never terminate at all.
+Consider a minor adaptation of the above example, where the author made a mistake in implementing the operator:
+
+```
+RECURSIVE MaxRec(_)
+MaxRec(seq) == IF seq = <<>>
+               THEN negInf
+               ELSE LET tailMax == MaxRec( seq ) \* forgot Tail!
+                    IN IF tailMax > Head(seq)
+                       THEN tailMax
+                       ELSE Head(seq)
+```
+
+Now, `MaxRec` never terminates, but spotting this error might not be trivial at a glance. This is where we believe folds hold the second advantage: `FoldSet` and `FoldSeq` *always terminate* in `Cardinality(set)` or `Len(seq)` steps, and each step is simple to describe, as it consists of a single operator application.
+
+In fact, the vast majority of the traditionally recursive operators can be equivalently rewritten as folds, for example:
+```
+RECURSIVE Cardinality(_)
+Cardinality(set) == IF set = {}
+                    THEN 0
+                    ELSE LET x == CHOOSE y \in set: TRUE
+                         IN 1 + Cardinality( set \ {x} )
+
+CardinalityFold(set) == LET Count(p,q) == p + 1 \* the value of q, the set element, is irrelevant
+                        IN FoldSet( Count, 0, set )
+```
+
+Notice that, in the case of sets, picking an arbitrary element `x`, to remove from the set at each step, utilizes the `CHOOSE` operator. This is a common trait shared by many operators that implement recursion over sets. 
+Since the introduction of folds, the use of `CHOOSE` in Apalache is heavily discouraged as it is both inefficient, as well as nondeterministic (unlike how `CHOOSE` is defined in TLA+ literature). 
```

Comment:
```suggestion
Since the introduction of folds, the use of `CHOOSE` in Apalache is heavily discouraged as it is both inefficient, as well as nondeterministic (unlike how `CHOOSE` is defined in TLA+ literature). For details, see the discussion in [issue 841](https://github.com/informalsystems/apalache/issues/841).
```